File: arena-junit/src/main/java/arena/junit/ClosedArena.java
package arena.junit;

import arena.junit.ffi.ArenaBindingError;
import arena.junit.ffi.ArenaBindings;
import arena.junit.ffi.ArenaLogLevel;
import arena.junit.match.Match;
import arena.junit.support.ArenaJson;
import com.sun.jna.Pointer;
import java.util.List;
import org.slf4j.ILoggerFactory;
import org.slf4j.Logger;

public final class ClosedArena {
  private final String name;
  private final List<Match> matches;
  private final ArenaLogLevel logLevel;
  private final Logger slf4jLogger;
  private final ILoggerFactory slf4jLoggerFactory;
  private final List<String> logDependencyIds;
  private final List<String> logComponentIds;

  public ClosedArena(String name, List<Match> matches) {
    this(name, matches, ArenaLogLevel.INFO, null, List.of(), List.of());
  }

  public ClosedArena(String name, List<Match> matches, ArenaLogLevel logLevel) {
    this(name, matches, logLevel, null, List.of(), List.of());
  }

  public ClosedArena(String name, List<Match> matches, ArenaLogLevel logLevel, Logger slf4jLogger) {
    this(name, matches, logLevel, slf4jLogger, List.of(), List.of());
  }

  public ClosedArena(
      String name,
      List<Match> matches,
      ArenaLogLevel logLevel,
      ILoggerFactory slf4jLoggerFactory) {
    this(name, matches, logLevel, null, slf4jLoggerFactory, List.of(), List.of());
  }

  public ClosedArena(
      String name,
      List<Match> matches,
      ArenaLogLevel logLevel,
      Logger slf4jLogger,
      List<String> logDependencyIds,
      List<String> logComponentIds) {
    this(name, matches, logLevel, slf4jLogger, null, logDependencyIds, logComponentIds);
  }

  public ClosedArena(
      String name,
      List<Match> matches,
      ArenaLogLevel logLevel,
      Logger slf4jLogger,
      ILoggerFactory slf4jLoggerFactory,
      List<String> logDependencyIds,
      List<String> logComponentIds) {
    this.name = name;
    this.matches = List.copyOf(matches);
    this.logLevel = logLevel;
    this.slf4jLogger = slf4jLogger;
    this.slf4jLoggerFactory = slf4jLoggerFactory;
    this.logDependencyIds =
        logDependencyIds == null || logDependencyIds.isEmpty()
            ? List.of()
            : List.copyOf(logDependencyIds);
    this.logComponentIds =
        logComponentIds == null || logComponentIds.isEmpty()
            ? List.of()
            : List.copyOf(logComponentIds);
  }

  private long registerLoggingTarget() {
    if (slf4jLoggerFactory != null) {
      return ArenaBindings.registerSlf4jDispatcherLoggingTarget(slf4jLoggerFactory, logLevel);
    }
    if (slf4jLogger != null) {
      return ArenaBindings.registerSlf4jDispatcherLoggingTarget(slf4jLogger, logLevel);
    }
    return ArenaBindings.registerDefaultDispatcherLoggingTarget(logLevel);
  }

  public OpenArena open() throws Exception {
    if (matches.isEmpty()) {
      throw new ArenaBindingError("closed arena has no matches");
    }
    String json = ArenaJson.MAPPER.writeValueAsString(matches.get(0).forFfi());
    ArenaBindings.setDispatcherDependencyAllowJson(
        logDependencyIds.isEmpty() ? null : ArenaJson.MAPPER.writeValueAsString(logDependencyIds));
    ArenaBindings.setDispatcherComponentAllowJson(
        logComponentIds.isEmpty() ? null : ArenaJson.MAPPER.writeValueAsString(logComponentIds));
    long logTok = registerLoggingTarget();
    try {
      Pointer h = ArenaBindings.arenaOpen(name, json, logLevel);
      return new OpenArena(h, logTok, matches);
    } catch (ArenaBindingError e) {
      ArenaBindings.unregisterDispatcherLoggingTarget(logTok);
      throw e;
    }
  }
}

File: arena-xunit/src/ClosedArena.cs
using System;
using System.Collections.Generic;
using System.Linq;
using ArenaDotnet.Xunit.Ffi;
using ArenaDotnet.Xunit.Playbook;
using ArenaDotnet.Xunit.Support;
using Microsoft.Extensions.Logging;

namespace ArenaDotnet.Xunit;

public sealed class ClosedArena
{
    private readonly string _name;
    private readonly Match _match;
    private readonly ArenaLogLevel _logLevel;
    private readonly ILogger? _logger;
    private readonly List<string>? _logDependencyIds;
    private readonly List<string>? _logComponentIds;

    public ClosedArena(string name, Match match)
        : this(name, match, ArenaLogLevel.Info, null, null, null)
    {
    }

    public ClosedArena(string name, Match match, ArenaLogLevel logLevel)
        : this(name, match, logLevel, null, null, null)
    {
    }

    public ClosedArena(string name, Match match, ArenaLogLevel logLevel, ILogger? logger)
        : this(name, match, logLevel, logger, null, null)
    {
    }

    public ClosedArena(string name, Match match, ArenaLogLevel logLevel, ILogger? logger,
        List<string>? logDependencyIds, List<string>? logComponentIds)
    {
        _name = name;
        _match = match;
        _logLevel = logLevel;
        _logger = logger;
        _logDependencyIds = logDependencyIds;
        _logComponentIds = logComponentIds;
    }

    public System.Threading.Tasks.Task<OpenArena> OpenAsync()
    {
        var json = _match.ForFfi();
        ArenaBindings.SetDispatcherDependencyAllowJson(
            _logDependencyIds != null && _logDependencyIds.Count > 0
                ? ArenaJson.Serialize(_logDependencyIds) : null);
        ArenaBindings.SetDispatcherComponentAllowJson(
            _logComponentIds != null && _logComponentIds.Count > 0
                ? ArenaJson.Serialize(_logComponentIds) : null);

        ulong logToken = _logger != null
            ? ArenaLogTarget.RegisterForLogger(_logger)
            : ArenaLogTarget.RegisterForLogger(CreateDefaultLogger());

        ArenaShutdown.EnsureHooksRegistered();

        IntPtr handle;
        try
        {
            handle = ArenaBindings.OpenArena(_name, json, _logLevel);
        }
        catch
        {
            ArenaLogTarget.Unregister(logToken);
            throw;
        }

        Dictionary<Type, ActivePlaybook> playbooks;
        try
        {
            playbooks = RunExecOnStartPlaybooks(handle);
        }
        catch
        {
            CloseArenaQuietly(handle);
            ArenaLogTarget.Unregister(logToken);
            throw;
        }

        return System.Threading.Tasks.Task.FromResult(new OpenArena(handle, logToken, _match, playbooks));
    }

    private static void CloseArenaQuietly(IntPtr handle)
    {
        try
        {
            ArenaBindings.CloseArena(handle);
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"arena: teardown after a failed open failed: {ex.Message}");
        }
    }

    private Dictionary<Type, ActivePlaybook> RunExecOnStartPlaybooks(IntPtr handle)
    {
        var result = new Dictionary<Type, ActivePlaybook>();
        try
        {
            foreach (var registered in _match.Playbooks)
            {
                if (registered.ExecOnDependencyStart)
                {
                    var playbookHandle = ArenaBindings.MatchPlaybookRun(handle, registered.Playbook.Identifier);
                    result[registered.Playbook.GetType()] = WrapActivePlaybook(registered.Playbook, playbookHandle);
                }
            }
        }
        catch
        {
            foreach (var started in result.Values)
            {
                try
                {
                    started.Dispose();
                }
                catch
                {
                }
            }
            throw;
        }
        return result;
    }

    private static ActivePlaybook WrapActivePlaybook(Playbook.IPlaybook playbook, IntPtr handle)
    {
        if (playbook is ManagedPlaybook managed)
            return managed.WrapHandle(handle);
        throw new InvalidOperationException($"unsupported playbook type: {playbook.GetType()}");
    }

    private static ILogger CreateDefaultLogger()
    {
        return new ConsoleLogger();
    }

    private sealed class ConsoleLogger : ILogger
    {
        public IDisposable? BeginScope<TState>(TState state) where TState : notnull => null;
        public bool IsEnabled(LogLevel level) => true;
        public void Log<TState>(LogLevel level, EventId eventId, TState state,
            Exception? exception, Func<TState, Exception?, string> formatter)
        {
            var msg = formatter(state, exception);
            Console.WriteLine($"{level}: {msg}");
        }
    }
}
